overlord/servicestate/quota_control.go

```diff
@@ -266,6 +292,38 @@ func UpdateQuota(st *state.State, name string, updateOpts QuotaGroupUpdate) (*st
 	return ts, nil
 }
 
+// remove a string item at index i from the string slice,
+// it maintains the ordering of the original slice.
+func remove(slice []string, i int) []string {
+	return append(slice[:i], slice[i+1:]...)
+}
+
+// ensureSnapServicesAbsentFromSubGroups removes all service references of a snap in
+// sub-groups related to the group of the snap, and returns the groups that were modified.
+func ensureSnapServicesAbsentFromSubGroups(grp *quota.Group, snap string, allGrps map[string]*quota.Group) ([]*quota.Group, error) {
+	// We can assume that when removing a snap, that if it has services in
+	// sub-groups that there will be only one level of sub-groups, so we can
+	// avoid checking further sub-groups.
+	var modifiedGrps []*quota.Group
+	for _, name := range grp.SubGroups {
+		subgrp, ok := allGrps[name]
+		if !ok {
+			return nil, fmt.Errorf("non-existent sub-group %q", name)
+		}
+
+		for idx, svc := range subgrp.Services {
+			parts := strings.Split(svc, ".")
```

Comment:
Could there be more dots in the name? If you are removing the suffix, `strings.TrimSuffix(svc, ".service")` is more obvious; otherwise, please add a comment explaining what we are doing here.

Reply:
Ah, I think I got it now :-) You not only need to split the suffix, but the service name as well. So maybe you can do `strings.HasPrefix(svc, snap + ".")`. Still a 1-line comment might help :-)

Reply:
Ah good point, changed to `strings.HasPrefix(svc, snap + ".")`!